src/garage/torch/optimizers/conjugate_gradient_optimizer.py

```diff
@@ -0,0 +1,203 @@
+            if hx is None:
+                hvp[i] = torch.zeros_like(p)
+
+        flat_output = torch.cat([h.reshape(-1) for h in hvp])
+        return flat_output + reg_coeff * vector
+
+    return _eval
+
+
+def conjugate_gradient(f_Ax, b, cg_iters, residual_tol=1e-10):
+    """Use Conjugate Gradient iteration to solve Ax = b. Demmel p 312.
+
+    Args:
+        f_Ax (function): A function to compute Hessian vector product.
+        b (torch.Tensor): Right hand side of the equation to solve.
+        cg_iters (int): Number of iterations to run conjugate gradient
+            algorithm.
+        residual_tol (float): Tolerence for convergence.
+
+    Returns:
+        torch.Tensor: Solution x* for equation Ax = b.
+
+    """
+    p = b.clone()
+    r = b.clone()
+    x = torch.zeros_like(b)
+    rdotr = torch.dot(r, r)
+
+    for _ in range(cg_iters):
+        z = f_Ax(p)
+        v = rdotr / torch.dot(p, z)
+        x += v * p
+        r -= v * z
+        newrdotr = torch.dot(r, r)
+        mu = newrdotr / rdotr
+        p = r + mu * p
+
+        rdotr = newrdotr
+        if rdotr < residual_tol:
+            break
+    return x
+
+
+class ConjugateGradientOptimizer(Optimizer):
+    """Performs constrained optimization via backtracking line search.
+
+    The search direction is computed using a conjugate gradient algorithm,
+    which gives x = A^{-1}g, where A is a second order approximation of the
+    constraint and g is the gradient of the loss function.
+
+    Args:
+        params (iterable): Iterable of parameters to optimize.
+        max_constraint_value (float): Maximum constraint value.
+        cg_iters (int): The number of CG iterations used to calculate A^-1 g
+        max_backtracks (int): Max number of iterations for backtrack
+            linesearch.
+        backtrack_ratio (float): backtrack ratio for backtracking line search.
+        hvp_reg_coeff (float): A small value so that A -> A + reg*I. It is
+            used by Hessian Vector Product calculation.
+        accept_violation (bool): whether to accept the descent step if it
+            violates the line search condition after exhausting all
+            backtracking budgets.
+
+    """
+
+    def __init__(self,
+                 params,
+                 max_constraint_value,
+                 cg_iters=10,
+                 max_backtracks=15,
+                 backtrack_ratio=0.8,
+                 hvp_reg_coeff=1e-5,
+                 accept_violation=False):
+        super().__init__(params, {})
+        self._cg_iters = cg_iters
+        self._max_backtracks = max_backtracks
+        self._backtrack_ratio = backtrack_ratio
+        self._accept_violation = accept_violation
+        self._max_constraint_value = max_constraint_value
+        self._hvp_reg_coeff = hvp_reg_coeff
+
+    def step(self, closure):
+        """Take an optimization step.
+
+        Args:
+            closure (tuple[function]): Functions to compute loss and
```

Comment:
why not separate arguments?

Reply:
This is overriding `torch.optim.Optimizer` 's step function, which only takes one parameter.

Reply:
ah i see.

Reply:
it seems clear to me that torch intends `closure` to be a single function. If you want to pseudo-implement their API, I think it would be more analogous to change the amend the signature to accept a second closure (which computes the constraint).

Reply:
@krzentner